VGICP_SAM:

  # Topics
  pointCloudTopic_L: "ns2/velodyne_points"               # Point cloud data - LEFT
  pointCloudTopic_R: "ns1/velodyne_points"               # Point cloud data - RIGHT
  imuTopic: "imu/data_raw"                               # IMU data
  odomTopic: "odometry/imu"                              # IMU pre-preintegration odometry, same frequency as IMU
  gpsTopic: "gt/odom"                            # GPS odometry topic from navsat, see module_navsat.launch file

  # Frames
  lidarFrame: "base_link"
  baselinkFrame: "base_link"
  odometryFrame: "odom"
  mapFrame: "map"

  # GPS Settings
  useImuHeadingInitialization: true           # if using GPS data, set to "true"
  useGpsElevation: true                       # if GPS elevation is bad, set to "false"
  gpsCovThreshold: 0.1                       # m^2, threshold for using GPS data
  poseCovThreshold: 0.01                      # m^2, threshold for using GPS data
  loopClosureEnableFlag: true
  # Export settings
  savePCD: false
  savePCDDirectory: "/Downloads/LOAM/"        # in your home folder, starts and ends with "/". Warning: the code deletes "LOAM" folder then recreates it. See "mapOptimization" for implementation

  # Sensor Settings
  sensor: velodyne                            # lidar sensor type, either 'velodyne' or 'ouster'
  N_SCAN: 16                                  # number of lidar channel (i.e., 16, 32, 64, 128)
  Horizon_SCAN: 1800                          # lidar horizontal resolution (Velodyne:1800, Ouster:512,1024,2048)
  Angle_Resolution_X: 0.2;                    # Angle Resolution from VELODYNE  (X)
  Angle_Resolution_Y: 2.0;                    # Angle Resolution from VELODYNE  (Y)
  Angle_Bottom_Z : 15.1;                      # Angle Resolution from VELODYNE  (Z-Angle)

  downsampleRate: 1                           # default: 1. Downsample your data if too many points. i.e., 16 = 64 / 4, 16 = 16 / 1
  lidarMinRange: 0.0                          # default: 1.0, minimum lidar range to be used
  lidarMaxRange: 500.0                        # default: 1000.0, maximum lidar range to be used
  MultiLidarEnableFlag : 2                    # Using Multi-LiDAR Fusion / 0 for Single LiDAR
  MultiLidarOnlineCalibration : 0             # Online LiDAR Calibration for uncertainty Propagation
  useCloudRing : 0                            # Cloud Ring Feature

  # IMU Setting - Xsense MTI-30
  SENSOR_dt: 0
  IMU_HZ: 100
  imuAccNoise: 1.2596236087144755e-02
  imuGyrNoise: 1.0796013414777343e-02
  imuAccBiasN: 1.5023316483881740e-04
  imuGyrBiasN: 7.5884170969887886e-05

  imuGravity: 9.80511
  imuRPYWeight: 0.01

  # Extrinsics (VEHICLE -> IMU)
  extrinsicTrans: [-0.07, 0.0, 1.7]
  extrinsicRot: [1, 0, 0,
                  0, 1, 0,
                  0, 0, 1]
  extrinsicRPY: [1, 0, 0,
                  0, 1, 0,
                  0, 0, 1]


  # LOAM feature threshold
  edgeThreshold: 1.0
  surfThreshold: 0.1
  edgeFeatureMinValidNum: 10
  surfFeatureMinValidNum: 100

  # voxel filter paprams
  odometryLeafSize: 0.01                        # default: 0.4 - outdoor, 0.2 - indoor
  mappingCornerLeafSize: 1.0                    # default: 0.2 - outdoor, 0.1 - indoor
  mappingLeafSize: 0.2                          # default: 0.4 - outdoor, 0.2 - indoor - VGICP Voxel Grid Filter

  # robot motion constraint (in case you are using a 2D robot)
  z_tollerance: 1000                            # meters
  rotation_tollerance: 1000                     # radians

  # CPU Params
  numberOfCores: 12                              # number of cores for mapping optimization
  mappingProcessInterval: 0.01                  # seconds, regulate mapping frequency

  # Surrounding map
  surroundingkeyframeAddingDistThreshold: 0.05   # meters, regulate keyframe adding threshold
  surroundingkeyframeAddingAngleThreshold: 0.05  # radians, regulate keyframe adding threshold
  surroundingKeyframeDensity: 0.01               # meters, downsample surrounding keyframe poses - Odometry Leaf Size Purpose
  surroundingKeyframeSearchRadius: 50.0         # meters, within n meters scan-to-map optimization (when loop closure disabled)

  # Loop closure
  loopClosureFrequency: 10                     # Hz, regulate loop closure constraint add frequency
  surroundingKeyframeSize: 50                   # submap size (when loop closure enabled)
  historyKeyframeSearchRadius: 80.0             # meters, key frame that is within n meters from current pose will be considerd for loop closure
  historyKeyframeSearchTimeDiff: 30.0           # seconds, key frame that is n seconds older will be considered for loop closure
  historyKeyframeSearchNum: 20                  # number of hostory key frames will be fused into a submap for loop closure
  historyKeyframeFitnessScore: 0.3              # icp threshold, the smaller the better alignment

  # Visualization
  globalMapVisualizationSearchRadius: 1000.0    # meters, global map visualization radius
  globalMapVisualizationPoseDensity: 10.0       # meters, global map visualization keyframe density
  globalMapVisualizationLeafSize: 1.0           # meters, global map visualization cloud density

  # G-ICP Parameters
  TransformationEpsilon : 0.01
  Resolution : 0.7
  MaximumIterations : 60
  MaxCorrespondenceDistance : 1.0

# Navsat (convert GPS coordinates to Cartesian)
navsat:
  frequency: 50
  wait_for_datum: false
  delay: 0.0
  magnetic_declination_radians: 0
  yaw_offset: 0
  zero_altitude: true
  broadcast_utm_transform: false
  broadcast_utm_transform_as_parent_frame: true
  publish_filtered_gps: false

# EKF for Navsat
ekf_gps:
  publish_tf: false
  map_frame: map
  odom_frame: odom
  base_link_frame: base_link
  world_frame: odom

  frequency: 50
  two_d_mode: false
  sensor_timeout: 0.01
  # -------------------------------------
  # External IMU:
  # -------------------------------------
  imu0: imu/data_raw
  # make sure the input is aligned with ROS REP105. "imu_correct" is manually transformed by myself. EKF can also transform the data using tf between your imu and base_link
  imu0_config: [false, false, false,
                true,  true,  true,
                false, false, false,
                false, false, true,
                true,  true,  true]
  imu0_differential: false
  imu0_queue_size: 50
  imu0_remove_gravitational_acceleration: true
  # -------------------------------------
  # Odometry (From Navsat):
  # -------------------------------------
  odom0: odometry/gps
  odom0_config: [true,  true,  true,
                 false, false, false,
                 false, false, false,
                 false, false, false,
                 false, false, false]
  odom0_differential: false
  odom0_queue_size: 10

  #                            x     y     z     r     p     y   x_dot  y_dot  z_dot  r_dot p_dot y_dot x_ddot y_ddot z_ddot
  process_noise_covariance: [  1.0,  0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    1.0,  0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    10.0, 0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0.03, 0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0.03, 0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0.1,  0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0.25,  0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0.25,  0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0.04,  0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0.01, 0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0.01, 0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0.5,  0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0.01, 0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0.01,   0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0.015]
